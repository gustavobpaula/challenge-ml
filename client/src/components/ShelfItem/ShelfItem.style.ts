import { css } from '@emotion/react'
import styled from '@emotion/styled'

export const ShelfItem = styled.article`
  display: flex;
`

export const Picture = styled.figure`
  display: inline-table;
  justify-content: center;
  margin: ${({ theme }) => theme.spacings.rule1};
  margin-left: 0;

  ${({ theme }) => css`
    @media (min-width: ${theme.breakpoints.forDesktopUp}) {
      display: flex;
      height: 180px;
      width: 180px;
    }
  `}

  width: 120px;

  img {
    border-radius: 4px;
    max-height: 100%;
    max-width: 100%;
  }
`

export const Description = styled.div`
  display: flex;
  flex-direction: column;
  margin-top: ${({ theme }) => theme.spacings.rule1};
`

export const Price = styled.div`
  margin-bottom: ${({ theme }) => theme.spacings.rule3};
  margin-top: ${({ theme }) => theme.spacings.rule1};
`

export const Value = styled.h3`
  color: ${({ theme }) => theme.colors.gray1};
  display: inline-block;
  font-size: ${({ theme }) => theme.type.sizes.lg};
  font-weight: normal;
  margin: 0;
`

export const Title = styled.h2`
  color: ${({ theme }) => theme.colors.gray1};
  font-size: ${({ theme }) => theme.type.sizes.md};
  font-weight: normal;
  margin: 0;
`

export const FreeShipping = styled.img`
  margin-left: ${({ theme }) => theme.spacings.rule1};
`

export const Locate = styled.p`
  ${({ theme }) => css`
    color: ${theme.colors.gray2};
    display: none;
    font-size: ${theme.type.sizes.xxs};
    margin-left: auto;
    margin-right: ${theme.spacings.rule3};
    margin-top: ${theme.spacings.rule3};

    @media (min-width: ${theme.breakpoints.forDesktopUp}) {
      display: block;
    }
  `}
`
