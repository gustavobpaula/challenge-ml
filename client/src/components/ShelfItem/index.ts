export { default as ShelfItem } from './ShelfItem'
