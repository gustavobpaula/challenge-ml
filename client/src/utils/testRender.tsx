import React from 'react'
import { render, RenderOptions } from '@testing-library/react'
import { ThemeProvider } from '@emotion/react'

import theme from 'config/theme'

type RenderComponentProps = {
  component: React.ReactNode
  useTheme?: boolean
  options?: RenderOptions
}

type ProviderProps = {
  children: React.ReactNode
}

const Theme = ({ children }: ProviderProps) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
)

export const RenderComponent = ({
  component,
  useTheme = false,
  options = {},
}: RenderComponentProps) => {
  const providers = [useTheme && Theme]

  const componentWithProviders = providers.reduce((accComponent, Provider) => {
    if (!Provider) return accComponent

    return <Provider>{accComponent}</Provider>
  }, component)

  return render(<>{componentWithProviders}</>, options)
}
