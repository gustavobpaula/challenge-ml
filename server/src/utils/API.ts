import fetch from 'node-fetch'
import { URLSearchParams } from 'url'

const ML_API = 'https://api.mercadolibre.com'

const API = {
  /**
   * Return currency object
   */
  getCurrency: async (id: string) => {
    try {
      const response = await fetch(`${ML_API}/currencies/${id}`)
      return await response.json()
    } catch (error) {
      throw new Error(error)
    }
  },
  /**
   * Return items by query
   */
  search: async (query: string) => {
    const params = new URLSearchParams()
    params.append('q', query)

    try {
      const response = await fetch(
        `${ML_API}/sites/MLA/search?${params.toString()}`
      )

      return await response.json()
    } catch (error) {
      throw new Error(error)
    }
  },
}

export default API
