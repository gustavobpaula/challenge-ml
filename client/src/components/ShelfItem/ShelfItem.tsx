import { Link } from 'react-router-dom'
import { ItemProps } from 'typings/SearchResultProps'
import truck from 'assets/ic_shipping.png'

import * as S from './ShelfItem.style'

type ShelfItemProps = {
  item: ItemProps
}

function ShelfItem({ item }: ShelfItemProps) {
  const urlToRedirect = `/items/${item.id}`

  return (
    <S.ShelfItem>
      <Link to={urlToRedirect}>
        <S.Picture>
          <img src={item.picture} loading="lazy" alt={item.title} />
        </S.Picture>
      </Link>

      <S.Description>
        <S.Price>
          <Link to={urlToRedirect}>
            <S.Value>
              {item.price.currency} {item.price.amount?.toLocaleString('es-AR')}
            </S.Value>
          </Link>
          {item.free_shipping && (
            <S.FreeShipping src={truck} alt="free shipping" />
          )}
        </S.Price>
        <Link to={urlToRedirect}>
          <S.Title>{item.title}</S.Title>
        </Link>
      </S.Description>
      <S.Locate>{item.state}</S.Locate>
    </S.ShelfItem>
  )
}

export default ShelfItem
