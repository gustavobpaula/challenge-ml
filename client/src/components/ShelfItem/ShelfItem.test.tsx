import { screen } from '@testing-library/react'

import { RenderComponent } from 'utils'
import ShelfItem, { ShelfItemProps } from './ShelfItem'
import productItems from '__mocks__/items'

const makeSut = (overWriteProps?: Partial<ShelfItemProps>) => {
  const defaultProps: ShelfItemProps = {
    item: productItems[0],
  }
  return RenderComponent({
    component: <ShelfItem {...defaultProps} {...overWriteProps} />,
    useTheme: true,
    useRouter: true,
  })
}

describe('ShelfItem component', () => {
  it('Should render match snapshot', () => {
    const { asFragment } = makeSut()

    expect(asFragment()).toMatchSnapshot()
  })

  it('Should render a image', () => {
    const product = productItems[0]
    makeSut({ item: product })

    expect(screen.getByRole('img', { name: product.title })).toBeInTheDocument()
  })

  it('Should render a price formatted', () => {
    const product = productItems[0]
    makeSut({ item: product })

    const price = `${
      product.price.currency
    } ${product.price.amount.toLocaleString('es-AR')}`

    expect(screen.queryByText(price)).toBeInTheDocument()
  })

  it('Should render a product title', () => {
    const product = productItems[0]
    makeSut({ item: product })

    expect(screen.queryByText(product.title)).toBeInTheDocument()
  })

  it('Should render a state', () => {
    const product = productItems[0]
    makeSut({ item: product })

    const { state = '' } = product

    expect(screen.queryByText(state)).toBeInTheDocument()
  })
})
