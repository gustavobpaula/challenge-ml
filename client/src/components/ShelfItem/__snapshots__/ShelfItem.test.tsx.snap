// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShelfItem component Should render match snapshot 1`] = `
<DocumentFragment>
  <article
    class="css-ov1ktg"
  >
    <a
      href="/items/MLA916188006"
    >
      <figure
        class="css-qy8l6h"
      >
        <img
          alt="iPod Touch 32gb - Blue"
          loading="lazy"
          src="http://http2.mlstatic.com/D_961656-MLA45566819084_042021-O.jpg"
        />
      </figure>
    </a>
    <div
      class="css-1gen4p5"
    >
      <div
        class="css-1hulj7l"
      >
        <a
          href="/items/MLA916188006"
        >
          <h3
            class="css-1a9qawe"
          >
            $ 39.999
          </h3>
        </a>
        <img
          alt="free shipping"
          class="css-ci461p"
          src="ic_shipping.png"
        />
      </div>
      <a
        href="/items/MLA916188006"
      >
        <h2
          class="css-vv2ccr"
        >
          iPod Touch 32gb - Blue
        </h2>
      </a>
    </div>
    <p
      class="css-1nktgjh"
    >
      Capital Federal
    </p>
  </article>
</DocumentFragment>
`;
